File: src/chessie/core/notation/san.py
"""SAN (Standard Algebraic Notation) conversion and parsing."""

from __future__ import annotations

from chessie.core.enums import MoveFlag, PieceType
from chessie.core.move import Move
from chessie.core.move_generator import MoveGenerator
from chessie.core.position import Position
from chessie.core.types import file_of, parse_square, rank_of, square_name

_SAN_PIECE: dict[PieceType, str] = {
    PieceType.KNIGHT: "N",
    PieceType.BISHOP: "B",
    PieceType.ROOK: "R",
    PieceType.QUEEN: "Q",
    PieceType.KING: "K",
}
_SAN_PIECE_REV: dict[str, PieceType] = {v: k for k, v in _SAN_PIECE.items()}


def move_to_san(position: Position, move: Move) -> str:
    """Convert a legal *move* to SAN given the *position* before the move."""
    board = position.board
    piece = board[move.from_sq]
    assert piece is not None

    # Castling
    if move.flag == MoveFlag.CASTLE_KINGSIDE:
        san = "O-O"
    elif move.flag == MoveFlag.CASTLE_QUEENSIDE:
        san = "O-O-O"
    else:
        san = ""
        is_capture = board[move.to_sq] is not None or move.flag == MoveFlag.EN_PASSANT

        if piece.piece_type == PieceType.PAWN:
            if is_capture:
                san += chr(ord("a") + file_of(move.from_sq))
        else:
            san += _SAN_PIECE[piece.piece_type]

            # Disambiguation
            gen = MoveGenerator(position)
            legal = gen.generate_legal_moves()
            ambiguous = [
                m
                for m in legal
                if m.to_sq == move.to_sq
                and m.from_sq != move.from_sq
                and board[m.from_sq] is not None
                and board[m.from_sq].piece_type == piece.piece_type  # type: ignore[union-attr]
            ]
            if ambiguous:
                same_file = any(
                    file_of(m.from_sq) == file_of(move.from_sq) for m in ambiguous
                )
                same_rank = any(
                    rank_of(m.from_sq) == rank_of(move.from_sq) for m in ambiguous
                )
                if not same_file:
                    san += chr(ord("a") + file_of(move.from_sq))
                elif not same_rank:
                    san += str(rank_of(move.from_sq) + 1)
                else:
                    san += square_name(move.from_sq)

        if is_capture:
            san += "x"

        san += square_name(move.to_sq)

        if move.flag == MoveFlag.PROMOTION and move.promotion is not None:
            san += "=" + _SAN_PIECE[move.promotion]

    # Check / checkmate suffix
    position.make_move(move)
    gen_after = MoveGenerator(position)
    if gen_after.is_in_check(position.side_to_move):
        legal_after = gen_after.generate_legal_moves()
        san += "#" if not legal_after else "+"
    position.unmake_move(move)

    return san


def parse_san(position: Position, san: str) -> Move:
    """Parse a SAN string into a :class:`Move` given the current *position*."""
    gen = MoveGenerator(position)
    legal = gen.generate_legal_moves()

    clean = san.rstrip("+#!?")

    # Castling
    if clean in ("O-O", "0-0"):
        for m in legal:
            if m.flag == MoveFlag.CASTLE_KINGSIDE:
                return m
        raise ValueError(f"Illegal move: {san}")

    if clean in ("O-O-O", "0-0-0"):
        for m in legal:
            if m.flag == MoveFlag.CASTLE_QUEENSIDE:
                return m
        raise ValueError(f"Illegal move: {san}")

    # Promotion
    promotion: PieceType | None = None
    if "=" in clean:
        promotion = _SAN_PIECE_REV.get(clean[-1])
        clean = clean[:-2]  # drop "=Q" etc.

    # Destination (last two chars)
    to_sq = parse_square(clean[-2:])
    clean = clean[:-2]

    # Capture marker
    if clean.endswith("x"):
        clean = clean[:-1]

    # Piece type
    if clean and clean[0] in _SAN_PIECE_REV:
        piece_type = _SAN_PIECE_REV[clean[0]]
        clean = clean[1:]
    else:
        piece_type = PieceType.PAWN

    # Disambiguation
    from_file: int | None = None
    from_rank: int | None = None
    if len(clean) == 2:
        from_file = ord(clean[0]) - ord("a")
        from_rank = int(clean[1]) - 1
    elif len(clean) == 1:
        if clean[0].isalpha():
            from_file = ord(clean[0]) - ord("a")
        else:
            from_rank = int(clean[0]) - 1

    # Find matching legal move
    candidates: list[Move] = []
    for m in legal:
        p = position.board[m.from_sq]
        if p is None or p.piece_type != piece_type:
            continue
        if m.to_sq != to_sq:
            continue
        if promotion is not None and m.promotion != promotion:
            continue
        if from_file is not None and file_of(m.from_sq) != from_file:
            continue
        if from_rank is not None and rank_of(m.from_sq) != from_rank:
            continue
        candidates.append(m)

    if len(candidates) == 1:
        return candidates[0]
    if not candidates:
        raise ValueError(f"Illegal move: {san}")
    raise ValueError(f"Ambiguous move: {san} → {candidates}")

File: src/notation/san.cpp
#include <chessie/notation/san.hpp>

#include <chessie/movegen.hpp>

#include <cctype>
#include <stdexcept>
#include <unordered_map>

namespace chessie {

namespace {

const std::unordered_map<PieceType, char> kSanPiece = {
    {PieceType::Knight, 'N'},
    {PieceType::Bishop, 'B'},
    {PieceType::Rook, 'R'},
    {PieceType::Queen, 'Q'},
    {PieceType::King, 'K'},
};

const std::unordered_map<char, PieceType> kSanPieceRev = {
    {'N', PieceType::Knight},
    {'B', PieceType::Bishop},
    {'R', PieceType::Rook},
    {'Q', PieceType::Queen},
    {'K', PieceType::King},
};

}  // namespace

std::string move_to_san(Position& position, Move move) {
    const Board& board = position.board();
    const Piece piece = board.piece_at(move.from_sq);
    if (piece.type == PieceType::None) {
        throw std::invalid_argument("move_to_san: empty from-square");
    }

    std::string san;
    if (move.flag == MoveFlag::CastleKingside) {
        san = "O-O";
    } else if (move.flag == MoveFlag::CastleQueenside) {
        san = "O-O-O";
    } else {
        const bool is_capture =
            board.piece_at(move.to_sq).type != PieceType::None || move.flag == MoveFlag::EnPassant;

        if (piece.type == PieceType::Pawn) {
            if (is_capture) {
                san += static_cast<char>('a' + file_of(move.from_sq));
            }
        } else {
            san += kSanPiece.at(piece.type);

            const MoveList legal = movegen::legal(position);
            std::vector<Move> ambiguous;
            for (const Move& m : legal) {
                if (m.to_sq != move.to_sq || m.from_sq == move.from_sq) {
                    continue;
                }
                const Piece other = board.piece_at(m.from_sq);
                if (other.type == piece.type) {
                    ambiguous.push_back(m);
                }
            }

            if (!ambiguous.empty()) {
                bool same_file = false;
                bool same_rank = false;
                for (const Move& m : ambiguous) {
                    if (file_of(m.from_sq) == file_of(move.from_sq)) {
                        same_file = true;
                    }
                    if (rank_of(m.from_sq) == rank_of(move.from_sq)) {
                        same_rank = true;
                    }
                }
                if (!same_file) {
                    san += static_cast<char>('a' + file_of(move.from_sq));
                } else if (!same_rank) {
                    san += static_cast<char>('1' + rank_of(move.from_sq));
                } else {
                    san += square_name(move.from_sq);
                }
            }
        }

        if (is_capture) {
            san += 'x';
        }
        san += square_name(move.to_sq);

        if (move.flag == MoveFlag::Promotion && move.promotion != PieceType::None) {
            san += '=';
            san += kSanPiece.at(move.promotion);
        }
    }

    position.make_move(move);
    if (position.is_in_check(position.side_to_move())) {
        const MoveList legal_after = movegen::legal(position);
        san += legal_after.empty() ? '#' : '+';
    }
    position.unmake_move(move);

    return san;
}

Move parse_san(Position& position, std::string_view san) {
    const MoveList legal = movegen::legal(position);

    std::string clean(san);
    while (!clean.empty() && (clean.back() == '+' || clean.back() == '#' || clean.back() == '!' ||
                              clean.back() == '?')) {
        clean.pop_back();
    }

    if (clean == "O-O" || clean == "0-0") {
        for (const Move& m : legal) {
            if (m.flag == MoveFlag::CastleKingside) {
                return m;
            }
        }
        throw std::invalid_argument(std::string("Illegal move: ") + std::string(san));
    }

    if (clean == "O-O-O" || clean == "0-0-0") {
        for (const Move& m : legal) {
            if (m.flag == MoveFlag::CastleQueenside) {
                return m;
            }
        }
        throw std::invalid_argument(std::string("Illegal move: ") + std::string(san));
    }

    PieceType promotion = PieceType::None;
    if (const auto eq = clean.find('='); eq != std::string::npos) {
        const char promo_char = clean.back();
        const auto it = kSanPieceRev.find(promo_char);
        if (it == kSanPieceRev.end()) {
            throw std::invalid_argument(std::string("Illegal move: ") + std::string(san));
        }
        promotion = it->second;
        clean.erase(eq);
    }

    if (clean.size() < 2) {
        throw std::invalid_argument(std::string("Illegal move: ") + std::string(san));
    }

    const Square to_sq = parse_square(clean.substr(clean.size() - 2));
    if (to_sq == kNoSquare) {
        throw std::invalid_argument(std::string("Illegal move: ") + std::string(san));
    }
    clean.erase(clean.size() - 2);

    if (!clean.empty() && clean.back() == 'x') {
        clean.pop_back();
    }

    PieceType piece_type = PieceType::Pawn;
    if (!clean.empty() && kSanPieceRev.contains(clean.front())) {
        piece_type = kSanPieceRev.at(clean.front());
        clean.erase(clean.begin());
    }

    std::optional<int> from_file;
    std::optional<int> from_rank;
    if (clean.size() == 2) {
        from_file = clean[0] - 'a';
        from_rank = clean[1] - '1';
    } else if (clean.size() == 1) {
        if (std::isalpha(static_cast<unsigned char>(clean[0])) != 0) {
            from_file = clean[0] - 'a';
        } else {
            from_rank = clean[0] - '1';
        }
    }

    std::vector<Move> candidates;
    for (const Move& m : legal) {
        const Piece p = position.board().piece_at(m.from_sq);
        if (p.type != piece_type) {
            continue;
        }
        if (m.to_sq != to_sq) {
            continue;
        }
        if (promotion != PieceType::None && m.promotion != promotion) {
            continue;
        }
        if (from_file.has_value() && file_of(m.from_sq) != *from_file) {
            continue;
        }
        if (from_rank.has_value() && rank_of(m.from_sq) != *from_rank) {
            continue;
        }
        candidates.push_back(m);
    }

    if (candidates.size() == 1) {
        return candidates.front();
    }
    if (candidates.empty()) {
        throw std::invalid_argument(std::string("Illegal move: ") + std::string(san));
    }
    throw std::invalid_argument(std::string("Ambiguous move: ") + std::string(san));
}

}  // namespace chessie
